Source/Core/Common/ucrtFreadWorkaround.cpp

```diff
@@ -4,93 +4,122 @@
+    qfe = static_cast<u16>(rhs);
+    return *this;
+  }
 };
 
-bool ApplyPatch(const PatchInfo& patch)
+static bool GetModulePath(const wchar_t *name, std::wstring *path)
 {
-  auto module = GetModuleHandleW(patch.module_name);
-  if (module == nullptr)
-  {
+  auto module = GetModuleHandleW(name);
+  if (module == nullptr) {
     return false;
   }
-
-  auto ucrtbase_pe = (PIMAGE_NT_HEADERS)((uintptr_t)module + ((PIMAGE_DOS_HEADER)module)->e_lfanew);
-  if (ucrtbase_pe->OptionalHeader.CheckSum != patch.checksum)
-  {
+  DWORD path_len = MAX_PATH;
+retry:
+  path->resize(path_len);
+  path_len = GetModuleFileNameW(module, const_cast<wchar_t *>(path->data()),
+    static_cast<DWORD>(path->size()));
+  if (!path_len) {
     return false;
   }
+  auto error = GetLastError();
+  if (error == ERROR_SUCCESS) {
+    return true;
+  }
+  if (error == ERROR_INSUFFICIENT_BUFFER) {
+    goto retry;
+  }
+  return false;
+}
 
-  void* patch_addr = (void*)((uintptr_t)module + patch.rva);
-  size_t patch_size = patch.length;
-
-  DWORD old_protect;
-  if (!VirtualProtect(patch_addr, patch_size, PAGE_EXECUTE_READWRITE, &old_protect))
-  {
+static bool GetModuleVersion(const wchar_t *name, Version *version)
+{
+  std::wstring path;
+  if (!GetModulePath(name, &path)) {
     return false;
   }
 
-  memset(patch_addr, 0x90, patch_size);
-
-  VirtualProtect(patch_addr, patch_size, old_protect, &old_protect);
-
-  FlushInstructionCache(GetCurrentProcess(), patch_addr, patch_size);
-
+  bool rv = false;
+  DWORD handle;
+  DWORD data_len = GetFileVersionInfoSizeW(path.c_str(), &handle);
+  if (!data_len) {
+    return false;
+  }
+  std::vector<u8> block(data_len);
+  if (!GetFileVersionInfoW(path.c_str(), handle, data_len, block.data())) {
+    return false;
+  }
+  void *buf;
+  UINT buf_len;
+  if (!VerQueryValueW(block.data(), LR"(\)", &buf, &buf_len)) {
+    return false;
+  }
+  auto info = static_cast<VS_FIXEDFILEINFO *>(buf);
+  *version = (static_cast<u64>(info->dwFileVersionMS) << 32) | info->dwFileVersionLS;
   return true;
 }
 
-int __cdecl EnableucrtFreadWorkaround()
-{
-  // This patches ucrtbase such that fseek will always
-  // synchronize the file object's internal buffer.
-
-  bool applied_at_least_one = false;
-  for (const auto& patch : s_patches)
-  {
-    if (ApplyPatch(patch))
-    {
-      applied_at_least_one = true;
+int __cdecl EnableucrtFreadWorkaround() {
+  const wchar_t *const module_names[] = { L"ucrtbase.dll", L"ucrtbased.dll" };
+  for (const auto &name : module_names) {
+    Version version;
+    if (GetModuleVersion(name, &version)) {
+      if (version.build < fixed_build) {
+        auto msg = StringFromFormat("You are running %S version %d.%d.%d.%d.\n"
+          "An important fix affecting dolphin was introduced in build %d.\n"
+          "You can use dolphin, but there will be known bugs.\n"
+          "Please update this file by installing the latest UCRT:\n"
+          "<put_url_here>\n",
```

Comment:
This message needs to be translatable. (Using `GetStringT` is probably the best option if you aren't going to be using panic alerts, though I suppose you'll have to replace `MessageBoxA` with `MessageBoxW` and re-encode the string...)

Also, capital D in Dolphin, please.